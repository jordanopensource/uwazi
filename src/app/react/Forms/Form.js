/* eslint-disable max-classes-per-file */
//TODO: replace react redux form
import PropTypes from 'prop-types';
import { LocalForm as RRLF, Form as RRF } from 'react-redux-form';

class Form extends RRF {}

Form.propTypes = {
  children: PropTypes.node.isRequired,
};

class LocalForm extends RRLF {}

LocalForm.propTypes = {
  children: PropTypes.node.isRequired,
};
export { Form, LocalForm };
