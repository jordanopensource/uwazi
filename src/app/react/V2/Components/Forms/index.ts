export { InputField } from './InputField';
export { Select } from './Select';
export { MultiSelect } from './MultiSelect';
export { RadioSelect } from './RadioSelect';
export { ConfirmNavigationModal } from './ConfirmNavigationModal';
export { Label } from './Label';
export { Geolocation } from './Geolocation';
export { FileDropzone } from './FileDropzone';

export type { MultiSelectProps } from './MultiSelect';
export type { SelectProps, OptionSchema } from './Select';
export { Checkbox } from './Checkbox';
export { EnableButtonCheckbox } from './EnableButtonCheckbox';
export { ColorPicker } from './ColorPicker';
export { DatePicker, DateRangePicker } from './DatePicker/DatePicker';
export { MultiselectList } from './MultiselectList';
export type { MultiselectListOption } from './MultiselectList';
