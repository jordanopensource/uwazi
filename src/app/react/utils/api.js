import { redirect } from 'react-router-dom';
import { isClient } from 'app/utils';
import { notify } from 'app/Notifications/actions/notificationsActions';
import { store } from 'app/store';
import loadingBar from 'app/App/LoadingProgressBar';
import { t } from 'app/I18N';

import { APIURL } from '../config.js';
import request from '../../shared/JSONRequest';

let API_URL = APIURL;
let language;

const doneLoading = data => {
  loadingBar.done();
  return data;
};

const isNonUsualApiError = error => error.status && ![401, 404, 409, 500].includes(error.status);

const errorMessages = [
  {
    key: /PayloadTooLargeError: request entity too large/g,
    message: 'Request data too large. Please review text length in property values',
  },
  {
    key: /ERROR Failed to index documents: (.+)/g,
    message: 'Failed to index documents: {0} ',
  },
];

function extractMessageFromError(error) {
  let finalMessage = `An error has occurred${error.json.requestId ? `, it has been logged with request id #${error.json.requestId}.` : '.'}`;
  if (!error.json.error) return finalMessage;

  const errorMessage = errorMessages.find(errorExpression =>
    error.json.error.match(errorExpression.key)
  );
  if (errorMessage) {
    const matches = errorMessage.key.exec(error.json.error);
    finalMessage = errorMessage.message;
    for (let i = 0; i < matches.length - 1; i += 1) {
      finalMessage = finalMessage.replace(`{${i}}`, matches[1]);
    }
  }
  return finalMessage;
}

function extractMessageFromValidation(error) {
  if (!error.json.validations) return error.json.error;
  return error.json.validations.reduce((message, validationError, index) => {
    const separator = index < error.json.validations.length - 1 ? ',' : '';
    return `${message} ${validationError.instancePath} ${validationError.message}${separator}`;
  }, `${error.json.error}: `);
}

// eslint-disable-next-line max-statements
const handleErrorStatus = error => {
  let errorMessage;
  switch (error.status || true) {
    case 400:
    case 422: {
      errorMessage = extractMessageFromValidation(error);
      store.dispatch(
        notify(t('System', extractMessageFromValidation(error), null, false), 'danger')
      );
      break;
    }

    case 401:
      redirect('/login');
      break;

    case 404:
      redirect('/404');
      break;

    case 409: {
      errorMessage = error.json.error;
      store.dispatch(notify(t('System', error.json.error, null, false), 'warning'));
      break;
    }

    case 500: {
      errorMessage = extractMessageFromError(error);
      store.dispatch(notify(t('System', extractMessageFromError(error), null, false), 'danger'));
      break;
    }

    case isNonUsualApiError(error): {
      errorMessage = error.json.prettyMessage || error.json.error;
      store.dispatch(
        notify(t('System', error.json.prettyMessage || error.json.error, null, false), 'danger')
      );
      break;
    }

    case error instanceof TypeError: {
      errorMessage = 'Could not reach server. Please try again later.';
      store.dispatch(
        notify(
          t('System', 'Could not reach server. Please try again later.', null, false),
          'danger'
        )
      );
      break;
    }

    default: {
      errorMessage = 'An error occurred';
      store.dispatch(notify(t('System', 'An error occurred', null, false), 'danger'));
    }
  }

  return errorMessage;
};

const handleError = (e, endpoint) => {
  const error = e;
  error.endpoint = endpoint;

  if (!isClient) {
    return Promise.reject(error);
  }

  doneLoading();

  handleErrorStatus(error);
  return Promise.reject(error);
};

const _request = (url, req, method) => {
  const headers = {
    'Content-Language': language,
    ...req.headers,
    'X-Requested-With': 'XMLHttpRequest',
  };

  loadingBar.start();

  return request[method](API_URL + url, req.data, headers)
    .then(doneLoading)
    .catch(async e => {
      await handleError(e, { url, method });
    });
};

export default {
  get: (url, req = {}) => _request(url, req, 'get'),

  post: (url, req = {}) => _request(url, req, 'post'),

  put: (url, req = {}) => _request(url, req, 'put'),

  delete: (url, req = {}) => _request(url, req, 'delete'),

  cookie(c) {
    request.cookie(c);
  },

  locale(locale) {
    language = locale;
  },

  APIURL(url) {
    API_URL = url;
  },
};
