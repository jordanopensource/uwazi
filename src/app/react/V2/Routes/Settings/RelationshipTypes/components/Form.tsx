/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';

import { Translate } from 'app/I18N';
import { InputField } from 'app/V2/Components/Forms';
import { useForm } from 'react-hook-form';
import { Button, Card } from 'app/V2/Components/UI';
import { Relationships } from './TableComponents';

interface FormProps {
  closePanel: () => void;
  relationtype?: Relationships;
  submit: (formValues: Relationships) => void;
  currentTypes: Relationships[];
}

const Form = ({ closePanel, submit, relationtype, currentTypes }: FormProps) => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<Relationships>({
    values: relationtype,
    mode: 'onSubmit',
  });

  return (
    <div className="relative h-full">
      <form onSubmit={handleSubmit(submit)} id="relationship-type-form">
        <Card title={<Translate>Relationship Type</Translate>}>
          <div className="flex flex-col gap-4">
            <InputField
              id="relationship-type-name"
              data-testid="relationship-type-form-name"
              label={<Translate>Name</Translate>}
              {...register('name', {
                required: true,
                validate: {
                  alreadyExists: value => !currentTypes.some(type => type.name === value),
                },
              })}
              hasErrors={!!errors.name}
              errorMessage={
                errors.name?.type === 'alreadyExists' ? <Translate>Already exists</Translate> : null
              }
            />
          </div>
        </Card>
      </form>
      <div className="flex absolute bottom-0 gap-2 px-4 py-3 w-full">
        <Button
          styling="light"
          onClick={closePanel}
          className="grow"
          data-testid="relationship-type-form-cancel"
        >
          <Translate>Cancel</Translate>
        </Button>
        <Button
          className="grow"
          type="submit"
          form="relationship-type-form"
          data-testid="relationship-type-form-submit"
        >
          <Translate>Save</Translate>
        </Button>
      </div>
    </div>
  );
};

export { Form };
