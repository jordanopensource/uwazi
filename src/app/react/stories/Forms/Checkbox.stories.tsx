import React from 'react';
import { Meta, StoryObj } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { fn } from '@storybook/test';
import { Checkbox } from 'app/V2/Components/Forms';
import { LEGACY_createStore as createStore } from 'V2/shared/testingHelpers';
import { Provider } from 'react-redux';

const meta: Meta<typeof Checkbox> = {
  title: 'Forms/Checkbox',
  component: Checkbox,
  args: {
    onChange: fn(),
  },
};

type Story = StoryObj<typeof Checkbox>;

const Primary: Story = {
  render: args => (
    <Provider store={createStore()}>
      <Checkbox
        label={args.label}
        defaultChecked={args.defaultChecked}
        checked={args.checked}
        name={args.name}
        onChange={args.onChange}
        className={args.className}
        disabled={args.disabled}
      />
    </Provider>
  ),
};

const Basic: Story = {
  ...Primary,
  args: {
    label: 'Uwazi is awesome',
    name: 'awesomeness',
    defaultChecked: false,
    disabled: false,
    checked: false,
    className: '',
    onChange: action('changed'),
  },
};

export { Basic };

export default meta;
