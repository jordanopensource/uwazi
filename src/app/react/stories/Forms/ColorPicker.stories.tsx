import React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import { Provider } from 'jotai';
import { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';
import { action } from '@storybook/addon-actions';
import { ColorPicker } from 'app/V2/Components/Forms';
import { LEGACY_createStore as createStore, atomsGlobalState } from 'V2/shared/testingHelpers';

const meta: Meta<typeof ColorPicker> = {
  title: 'Forms/ColorPicker',
  component: ColorPicker,
  args: {
    onChange: fn(),
  },
  parameters: {
    actions: {
      handles: ['change'],
    },
  },
};

type Story = StoryObj<typeof ColorPicker>;

const Primary: Story = {
  render: args => (
    <ReduxProvider store={createStore()}>
      <Provider store={atomsGlobalState()}>
        <div className="tw-content" style={{ height: '300px' }}>
          <ColorPicker
            name={args.name}
            className={args.className}
            onChange={args.onChange}
            hasErrors={args.hasErrors}
          />
        </div>
      </Provider>
    </ReduxProvider>
  ),
};

const Basic: Story = {
  ...Primary,
  args: {
    name: 'color',
    className: '',
    hasErrors: false,
    onChange: action('changed'),
  },
};

export { Basic };

export default meta;
