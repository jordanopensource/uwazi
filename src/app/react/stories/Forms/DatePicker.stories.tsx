import React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import { Provider } from 'jotai';
import { Meta, StoryObj } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { fn } from '@storybook/test';
import { DatePicker } from 'app/V2/Components/Forms';
import { LEGACY_createStore as createStore, atomsGlobalState } from 'V2/shared/testingHelpers';

const meta: Meta<typeof DatePicker> = {
  title: 'Forms/DatePicker',
  component: DatePicker,
  args: {
    onChange: fn(),
    onBlur: fn(),
    clearFieldAction: fn(),
  },
  parameters: {
    actions: {
      handles: ['change'],
    },
  },
};

type Story = StoryObj<typeof DatePicker>;

const Primary: Story = {
  render: args => (
    <ReduxProvider store={createStore()}>
      <Provider store={atomsGlobalState()}>
        <DatePicker
          name={args.name}
          label={args.label}
          language={args.language}
          labelToday={args.labelToday}
          labelClear={args.labelClear}
          dateFormat={args.dateFormat}
          placeholder={args.placeholder}
          hideLabel={args.hideLabel}
          className={args.className}
          onChange={args.onChange}
          onBlur={args.onBlur}
          clearFieldAction={args.clearFieldAction}
        />
      </Provider>
    </ReduxProvider>
  ),
};

const Basic: Story = {
  ...Primary,
  args: {
    name: 'dateField',
    label: 'Fecha',
    language: 'es',
    dateFormat: 'dd-mm-yyyy',
    labelToday: 'Hoy',
    labelClear: 'Limpiar',
    placeholder: 'Seleccione una fecha',
    hideLabel: true,
    className: '',
    onChange: action('changed'),
    onBlur: action('blurred'),
    clearFieldAction: action('cleared'),
  },
};

export { Basic };

export default meta;
