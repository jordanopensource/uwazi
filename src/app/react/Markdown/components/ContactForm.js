import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { actions, Field } from 'react-redux-form';
import { FormGroup } from 'app/ReactReduxForms';
import { Translate, t } from 'app/I18N';
import api from 'app/utils/api';
import { Icon } from 'UI';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { notify } from 'app/Notifications/actions/notificationsActions';
import { RequestParams } from 'app/utils/RequestParams';
import { LocalForm } from 'app/Forms/Form';
import { FormCaptcha } from './FormCaptcha';

class ContactForm extends Component {
  constructor(props, context) {
    super(props, context);
    this.submit = this.submit.bind(this);
    this.validators = {
      name: { required: val => val && val.length > 2 },
      email: { required: val => val },
      message: { required: val => val && val.length > 4 },
      captcha: { required: val => val && val.text.length },
    };
    this.refreshFn = refresh => {
      this.refreshCaptcha = refresh;
    };
  }

  attachDispatch(dispatch) {
    this.formDispatch = dispatch;
  }

  resetForm() {
    this.formDispatch(actions.reset('contactform'));
  }

  async submit(values) {
    try {
      await api.post(
        'contact',
        new RequestParams({
          name: values.name,
          email: values.email,
          message: values.message,
          captcha: JSON.stringify(values.captcha),
        })
      );
      this.props.notify(t('System', 'Message sent', null, false), 'success');
      this.resetForm();
    } catch (e) {
      if (e.status === 403) {
        this.props.notify(e.json.error, 'danger');
      } else {
        this.props.notify(t('System', 'An error occurred', null, false), 'danger');
        this.resetForm();
      }
    }

    this.refreshCaptcha();
  }

  render() {
    return (
      <LocalForm
        validators={this.validators}
        model="contactform"
        className="contact-form"
        getDispatch={dispatch => this.attachDispatch(dispatch)}
        onSubmit={this.submit}
      >
        <FormGroup model=".name">
          <label className="form-group-label" htmlFor="name">
            <Translate>Name</Translate>
            <span className="required">*</span>
          </label>

          <Field model=".name">
            <input type="text" name="name" className="form-control" />
          </Field>
        </FormGroup>

        <FormGroup model=".email">
          <label className="form-group-label" htmlFor="email">
            <Translate>Email</Translate>
            <span className="required">*</span>
          </label>

          <Field model=".email">
            <input type="email" name="email" className="form-control" />
          </Field>
        </FormGroup>

        <FormGroup model=".message">
          <label className="form-group-label" htmlFor="message">
            <Translate>Message</Translate>
            <span className="required">*</span>
          </label>

          <Field model=".message">
            <textarea name="message" className="form-control" />
          </Field>
        </FormGroup>

        <FormCaptcha refresh={this.refreshFn} />

        <button type="submit" className="btn btn-success">
          <Icon icon="paper-plane" />
          &nbsp;
          <span className="btn-label">
            <Translate>Send</Translate>
          </span>
        </button>
      </LocalForm>
    );
  }
}

ContactForm.propTypes = {
  notify: PropTypes.func.isRequired,
};

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ notify }, dispatch);
}

export { ContactForm };
export default connect(null, mapDispatchToProps)(ContactForm);
