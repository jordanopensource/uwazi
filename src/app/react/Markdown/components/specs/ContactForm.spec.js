/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { LocalForm } from 'app/Forms/Form';
import { Captcha } from 'app/ReactReduxForms';
import api from 'app/utils/api';
import { renderConnectedMount } from 'app/utils/test/renderConnected';
import { ContactForm } from '../ContactForm.js';

describe('ContactForm', () => {
  let props;
  let instance;
  let component;
  const formValues = { name: 'test', email: 'test@test.com', message: 'test' };

  const prepareMocks = () => {
    instance = component.instance();
    instance.refreshCaptcha = jest.fn();
    instance.formDispatch = jest.fn();
    component.find(LocalForm).props().getDispatch(instance.formDispatch);
    const captcha = component.find(Captcha);
    captcha.props().refresh = instance.refreshCaptcha;
    captcha.props().refresh(instance.refreshCaptcha);
  };

  beforeEach(() => {
    spyOn(api, 'post').and.callFake(async () => Promise.resolve());
    props = { notify: () => {} };
  });

  it('should render the ContactForm', () => {
    const shallowRender = shallow(<ContactForm {...props} />);
    expect(shallowRender).toMatchSnapshot();
  });

  it('should submit the values and reset the form', async () => {
    component = renderConnectedMount(ContactForm, undefined, props, true);
    prepareMocks();

    await act(async () => {
      component
        .find(LocalForm)
        .props()
        .onSubmit({
          ...formValues,
          captcha: { id: '1', captcha: 'abcd' },
        });
    });

    expect(api.post).toHaveBeenCalledWith('contact', {
      data: {
        ...formValues,
        captcha: JSON.stringify({ id: '1', captcha: 'abcd' }),
      },
      headers: {},
    });

    expect(instance.refreshCaptcha).toHaveBeenCalled();
    expect(instance.formDispatch).toHaveBeenCalledWith({
      model: 'contactform',
      type: 'rrf/reset',
    });
  });
});
