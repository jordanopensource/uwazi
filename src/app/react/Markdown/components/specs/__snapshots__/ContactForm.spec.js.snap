// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm should render the ContactForm 1`] = `
<LocalForm
  className="contact-form"
  getDispatch={[Function]}
  initialState={Object {}}
  model="contactform"
  onSubmit={[Function]}
  validators={
    Object {
      "captcha": Object {
        "required": [Function],
      },
      "email": Object {
        "required": [Function],
      },
      "message": Object {
        "required": [Function],
      },
      "name": Object {
        "required": [Function],
      },
    }
  }
>
  <FormGroup
    model=".name"
  >
    <label
      className="form-group-label"
      htmlFor="name"
    >
      <Connect(Translate)>
        Name
      </Connect(Translate)>
      <span
        className="required"
      >
        *
      </span>
    </label>
    <Modeled(Connect(Field))
      model=".name"
    >
      <input
        className="form-control"
        name="name"
        type="text"
      />
    </Modeled(Connect(Field))>
  </FormGroup>
  <FormGroup
    model=".email"
  >
    <label
      className="form-group-label"
      htmlFor="email"
    >
      <Connect(Translate)>
        Email
      </Connect(Translate)>
      <span
        className="required"
      >
        *
      </span>
    </label>
    <Modeled(Connect(Field))
      model=".email"
    >
      <input
        className="form-control"
        name="email"
        type="email"
      />
    </Modeled(Connect(Field))>
  </FormGroup>
  <FormGroup
    model=".message"
  >
    <label
      className="form-group-label"
      htmlFor="message"
    >
      <Connect(Translate)>
        Message
      </Connect(Translate)>
      <span
        className="required"
      >
        *
      </span>
    </label>
    <Modeled(Connect(Field))
      model=".message"
    >
      <textarea
        className="form-control"
        name="message"
      />
    </Modeled(Connect(Field))>
  </FormGroup>
  <Memo(CaptchaComponent)
    refresh={[Function]}
  />
  <button
    className="btn btn-success"
    type="submit"
  >
    <Connect(Icon)
      icon="paper-plane"
    />
    Â 
    <span
      className="btn-label"
    >
      <Connect(Translate)>
        Send
      </Connect(Translate)>
    </span>
  </button>
</LocalForm>
`;
